### ISSUE TITLE
spread tolerance amounts across users

### ISSUE THREAD

**Issue by carchrae** [foodcoops/foodsoft#473]
we have our first successful order from foodsoft.  the co-op members are happy....  mostly.  :) 

we noticed that tolerance is not considered across the order articles.  so, one user may end up getting a lot of extra while another user may get only a small amount extra.  in our co-op we have a rule to add $5 worth of items to each member so that we can fill a case.  but we don't want any member to feel they are charged a lot more than another.

i would like to change it so that it spreads the amount of tolerance used across members.   

so, current system

user 1:  
* article A:  2+2 : 4 received (4 * $2 = 8 (2*$2 = $4 extra))
* article B:  3+5 : 8 received (8 * $3 = $24 (5*3 = $15 extra))
* total tolerance used: +7 ($19 extra)

user 2: 
* article A:  1+1 : 1 received (1 * $2 = $2 ($0 extra))
* article B:  9+4 : 10 received (10 * $3 = $30 ($3 extra))
* total tolerance used: +1 ($3 extra)

proposal to be more fair

user 1:  
* article A:  2+2 : 3 received (3 * $2 = 8 (1*$2 = $2 extra))
* article B:  3+5 : 8 received (8 * $3 = $24 (5*3 = $15 extra))
* total tolerance used: +6 ($17 extra)

user 2: 
* article A:  1+1 : 2 received (2 * $2 = $4 (1* $2 = $2 extra))
* article B:  9+4 : 13 received (13 * $3 = $39 ($15 extra))
* total tolerance used: +5 ($17 extra)

the goal would be to adjust tolerance to spread the cost.  i am sure there are other criteria that might be fair.  for example, spreading the tolerance based on the typical order amount of a user (eg, do not give a single person a lot of extra because they usually place small orders)

**Comment by carchrae** [comment]
ps, i will probably be working on this.  just noting it here in case someone already did it.  :)

**Comment by wvengen** [comment]
This is a feature that I've seen requested before, thanks for putting it down.
I think spreading the tolerance would be preferred in most cases.

**Comment by carchrae** [comment]
there is another fairness measure that i may include in this, which is 'who loses out when there is not enough to go around'.  

although i like the simplicity of the current allocation algorithm because it favours first-come, in our co-op, some people raised concerns when those ordering later didn't get anything but perhaps another person got 2 because they ordered earlier.  i realise this is exactly what tolerance is for, so i am going to work on educating our co-op about why the tolerance system is both easy to understand and, i think perhaps more fair than adjusting the minimum quantity a member orders.

the contentious part is why first-come wins, and it is arguable if this is always fair.  but it most likely easier to explain that an adaptive weighted random allocation algorithm....   :)   aside from simplicity, i also like that a co-op member can ensure they get at least the minimum of what they want by requesting half a case, then adding tolerance for the rest of the case.  first-come means we can tell them the minimum they'll get at the time they are ordering.  

anyway, this is just me rambling.  no need to comment on this.  i just mention it because it is another co-op issue of fairness.  i think in general, there is no easy answer to who 'who do we leave out when there is not enough'.

**Comment by carchrae** [comment]
i just fixed the values in the example above.  they didn't add up before.

**Comment by wvengen** [comment]
Yeah, I think Foodsoft could be improved by spreading tolerance over all members. If you _want_ items, you'd usually use quantity. If you're _ok_ to have more, that's what tolerance is for - and having much more than you really wanted may be a nuisance rather than a win, so I think minimizing the tolerance over all members would actually be better. Still, the distribution code is already a little complex, so the current solution may just suffice.

**Comment by carchrae** [comment]
our current solution is to allow the ordering person to adjust the allocations manually.  to aid them, i modified the show the total amount of tolerance used for each member, so they can see who is getting way more.  i show it expressed in $ and as a % of their order.  this page is very slow right now (i need to do better caching somewhere) 

a picture is probably easier to see

![image](https://cloud.githubusercontent.com/assets/959825/26702258/1bfb61f6-46d9-11e7-9d73-ea91f50fb80f.png)
 

`_article_single_goa.html.haml`

```haml
- percent_tolerance = ((goa.group_order.price ? goa.group_order.total_tolerance_cost/goa.group_order.price : 0)*100).round
- tolerance_available = goa.tolerance_price_limit-goa.tolerance_price
%tr{class: if goa.result == 0 then 'unavailable' end, id: "goa_#{goa.id}"}
  %td{:style => "width:70%"}
    %span= goa.group_order.ordergroup_name
    %span= "(#{number_to_currency(tolerance_available)} tolerance left)" if tolerance_available > 0
    %br
    %span.small Total tolerance cost #{percent_tolerance}% - (#{number_to_currency(goa.group_order.total_tolerance_cost)} of #{number_to_currency(goa.group_order.price)})
  %td.center= "#{goa.quantity} + #{goa.tolerance}"
  %td.center.input-delta= (edit or true rescue true) ? group_order_article_edit_result(goa) : goa.result
  %td.price{data: {value: goa.total_price}}= number_to_currency(goa.total_price)
```
